src/OrleansServiceFabricUtils/FabricGatewayProvider.cs

```diff
@@ -0,0 +1,166 @@
+        public async Task InitializeGatewayListProvider(ClientConfiguration clientConfiguration, Logger logger)
+        {
+            // TODO: inject these
+            var serviceName = new Uri(clientConfiguration.DataConnectionString);
+            var fabricClient = new FabricClient();
+            var queryManager = new FabricQueryManager(fabricClient, new ServicePartitionResolver(() => fabricClient));
+            this.fabricServiceSiloResolver = new FabricServiceSiloResolver(serviceName, queryManager, logger.GetLogger);
+            this.fabricServiceSiloResolver.Subscribe(this);
+
+            this.log = logger.GetLogger(nameof(FabricGatewayProvider));
+            await this.RefreshAsync();
+            this.timer = new Timer(this.Refresh, null, this.refreshPeriod, this.refreshPeriod);
+        }
+
+        /// <summary>
+        /// Returns the list of gateways (silos) that can be used by a client to connect to Orleans cluster.
+        /// </summary>
+        public Task<IList<Uri>> GetGateways() => Task.FromResult<IList<Uri>>(this.gateways);
+
+        /// <summary>
+        /// Specifies how often this IGatewayListProvider is refreshed, to have a bound on max staleness of its returned information.
+        /// </summary>
+        public TimeSpan MaxStaleness { get; }
+
+        /// <summary>
+        /// Specifies whether this IGatewayListProvider ever refreshes its returned information, or always returns the same gw list.
+        /// (currently only the static config based StaticGatewayListProvider is not updatable. All others are.)
+        /// </summary>
+        public bool IsUpdatable => true;
+
+        /// <summary>
+        /// Subscribes the provided <paramref name="subscriber"/> from notification events.
+        /// </summary>
+        /// <param name="subscriber">The listener.</param>
+        /// <returns>A value indicating whether the listener was subscribed.</returns>
+        public bool SubscribeToGatewayNotificationEvents(IGatewayListListener subscriber)
+        {
+            this.subscribers.TryAdd(subscriber, subscriber);
+            return true;
+        }
+
+        /// <summary>
+        /// Unsubscribes the provided <paramref name="listener"/> from notification events.
+        /// </summary>
+        /// <param name="listener">The listener.</param>
+        /// <returns>A value indicating whether the listener was unsubscribed.</returns>
+        public bool UnSubscribeFromGatewayNotificationEvents(IGatewayListListener listener)
+        {
+            this.subscribers.TryRemove(listener, out listener);
+            return true;
+        }
+
+        /// <summary>
+        /// Notifies this instance of an update to one or more partitions.
+        /// </summary>
+        /// <param name="silos">The updated set of partitions.</param>
+        public void OnUpdate(FabricSiloInfo[] silos)
+        {
+            this.gateways = silos.Select(silo => silo.GatewayAddress.ToGatewayUri()).ToList();
```

Comment:
>silos.Select(silo => silo.GatewayAddress.ToGatewayUri()).ToList(); [](start = 28, length = 66)

Doesn't this assume that every silo is a gateway? Or does it rely on cluster config being the same for all nodes?